pkg/targetallocator/adapters/config_to_prom_config.go

```diff
@@ -57,3 +68,208 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+	prometheus, err := ConfigToPromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusConfig, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	scrapeConfigsProperty, ok := prometheusConfig["scrape_configs"]
+	if !ok {
+		return nil, errorNoComponent("scrape_configs")
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
+
+	sdRegex := regexp.MustCompile(`^.*(sd|static)_configs$`)
+
+	for _, config := range scrapeConfigs {
+		scrapeConfig, ok := config.(map[interface{}]interface{})
+		if !ok {
+			return nil, errorNotAMap("scrape_config")
+		}
+
+		// Check for other types of service discovery configs (e.g. dns_sd_configs, file_sd_configs, etc.)
+		for key := range scrapeConfig {
+			keyStr, keyErr := key.(string)
+			if !keyErr {
+				continue
+			}
+			if sdRegex.MatchString(keyStr) {
+				delete(scrapeConfig, key)
+			}
+		}
+
+		jobNameProperty, ok := scrapeConfig["job_name"]
+		if !ok {
+			return nil, errorNotAString("job_name")
+		}
+
+		jobName, ok := jobNameProperty.(string)
+		if !ok {
+			return nil, errorNotAString("job_name is not a string")
+		}
+
+		escapedJob := url.QueryEscape(jobName)
+		scrapeConfig["http_sd_configs"] = []interface{}{
+			map[string]interface{}{
+				"url": fmt.Sprintf("http://%s:80/jobs/%s/targets?collector_id=$POD_NAME", taServiceName, escapedJob),
+			},
+		}
+	}
+
+	return prometheus, nil
+}
+
+// AddTAConfigToPromConfig adds or updates the target_allocator configuration in the Prometheus configuration.
+func AddTAConfigToPromConfig(cfg string, taServiceName string) (map[interface{}]interface{}, error) {
+	prometheus, err := ConfigToPromConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusCfg, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	// Create the TargetAllocConfig dynamically if it doesn't exist
+	if prometheus["target_allocator"] == nil {
+		prometheus["target_allocator"] = make(map[interface{}]interface{})
+	}
+
+	targetAllocatorCfg, ok := prometheus["target_allocator"].(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("target_allocator")
+	}
+
+	targetAllocatorCfg["endpoint"] = fmt.Sprintf("http://%s:80", taServiceName)
+	targetAllocatorCfg["interval"] = "30s"
+	targetAllocatorCfg["collector_id"] = "${POD_NAME}"
+
+	// we don't need the scrape configs here anymore with target allocator enabled
+	// Remove the scrape_configs key from the map
+	delete(prometheusCfg, "scrape_configs")
```

Comment:
Can you also add an explanation for this in the function comment?

Reply:
Done.